### PR TITLE
Feature/313 Transfer Fulfilment Duplicate Check

### PR DESCRIPTION
**Pull request by ggrg**
PR created at:
```
=============================== Coverage summary ===============================
Statements   : 94.61% ( 2968/3137 )
Branches     : 89.99% ( 692/769 )
Functions    : 97.92% ( 423/432 )
Lines        : 94.63% ( 2943/3110 )
================================================================================

1..1128
# tests 1128
# pass  1128

# ok
```
Current PR state:
```
=============================== Coverage summary ===============================
Statements   : 100% ( 3140/3140 )
Branches     : 100% ( 767/767 )
Functions    : 100% ( 432/432 )
Lines        : 100% ( 3113/3113 )
================================================================================

1..1182
# tests 1182
# pass  1182

# ok
```

### CODE REVIEW

File: src/handlers/transfers/handler.js [hunk 1]
```diff
@@ -39,30 +39,32 @@
...
-const errorModifiedReqDescription = Errors.getErrorDescription(errorModifiedReqCode)
-const errorInternalCode = 2001
-const errorInternalDescription = Errors.getErrorDescription(errorInternalCode)
-const errorTransferExpCode = 3300
-const errorTransferExpDescription = Errors.getErrorDescription(errorTransferExpCode)
-const errorTransferIdNotFoundCode = 3208
-const errorTransferIdNotFoundDescription = Errors.getErrorDescription(errorTransferIdNotFoundCode)
+const Uuid = require('uuid4')
+
+const eEnum = {
```
Comment (rmothilal) [inline src/handlers/transfers/handler.js]: would it not be better to store this enum in under lib/errors

File: src/handlers/transfers/handler.js [hunk 2]
```diff
@@ -39,30 +39,32 @@
...
 const Kafka = require('../lib/kafka')
 const Validator = require('./validator')
-const TransferState = require('../../lib/enum').TransferState
-const TransferStateEnum = require('../../lib/enum').TransferStateEnum
 const Enum = require('../../lib/enum')
-const TransferEventType = Enum.transferEventType
-const TransferEventAction = Enum.transferEventAction
+const TransferState = Enum.TransferState
+const TransferStateEnum = Enum.TransferStateEnum
+const TET = Enum.transferEventType
```
Comment (rmothilal) [inline src/handlers/transfers/handler.js]: For readability i would suggest changing this to transferEventTypeEnum or transferEventType

File: src/handlers/transfers/handler.js [hunk 3]
```diff
@@ -39,30 +39,32 @@
...
+const TEA = Enum.transferEventAction
```
Comment (rmothilal) [inline src/handlers/transfers/handler.js]: For readability i would suggest changing this to transferEventActionEnum or transferEventAction

File: src/handlers/transfers/handler.js [hunk 4]
```diff
@@ -103,141 +105,92 @@ const prepare = async (error, messages) => {
...
       histTimerEnd({ success: false, fspId: Config.INSTRUMENTATION_METRICS_LABELS.fspId })
       return true
     }
-    const payload = message.value.content.payload
-    const headers = message.value.content.headers
-
-    Logger.info('TransferService::prepare::checking for duplicates')
-    const { existsMatching, existsNotMatching } = await TransferService.validateDuplicateHash(payload)
+    // Action Letter (AL) is used to denote breadcrumb flow end
+    const AL = action === TEA.PREPARE ? 'P' : '?'
```
Comment (rmothilal) [inline src/handlers/transfers/handler.js]: For readability please rename to actionLetter also i would store 'P' and '?' under constants with a name that explains what it is

File: src/handlers/transfers/handler.js [hunk 5]
```diff
@@ -269,106 +226,144 @@ const fulfil = async (error, messages) => {
...
-    const transferId = message.value.id
-    const payload = message.value.content.payload
-    const headers = message.value.content.headers
-    if (metadata.event.type === TransferEventType.FULFIL &&
-      (metadata.event.action === TransferEventAction.COMMIT ||
-        metadata.event.action === TransferEventAction.REJECT ||
-        metadata.event.action === TransferEventAction.ABORT)) {
-      const existingTransfer = await TransferService.getById(transferId)
+    // Action Letter (AL) is used to denote breadcrumb flow end
+    const AL = action === TEA.COMMIT ? 'C' : (action === TEA.REJECT ? 'R' : (action === TEA.ABORT ? 'A' : '?'))
```
Comment (rmothilal) [inline src/handlers/transfers/handler.js]: For readability please rename to actionLetter also i would store 'R', 'A' and '?' under constants with a name that explains what it is or create an enum for all this lettering

File: src/handlers/transfers/handler.js [hunk 6]
```diff
@@ -410,71 +407,37 @@ const getTransfer = async (error, messages) => {
...
     }
-    if (!await Validator.validateParticipantByName(message.value.from)) {
-      Logger.info('TransferService::getTransferHandler::participantCheck::doesntExist:: FSP Id does not exist')
-      if (!Kafka.Consumer.isConsumerAutoCommitEnabled(kafkaTopic)) {
-        await consumer.commitMessageSync(message)
-      }
-      histTimerEnd({ success: true, fspId: Config.INSTRUMENTATION_METRICS_LABELS.fspId })
-      return true
-    }
+    const AL = 'G'
```
Comment (rmothilal) [inline src/handlers/transfers/handler.js]: For readability please rename to actionLetter also i would store 'G' under a constant or enum with a name that explains what it is

**Comment by rmothilal** [comment]
also your test coverage seems to be failing

File: src/handlers/transfers/handler.js (hunk 1 above)
Comment (ggrg) [reply src/handlers/transfers/handler.js]: ok

File: src/handlers/transfers/handler.js (hunk 2 above)
Comment (ggrg) [reply src/handlers/transfers/handler.js]: ok

File: src/handlers/transfers/handler.js (hunk 3 above)
Comment (ggrg) [reply src/handlers/transfers/handler.js]: ok

File: src/handlers/transfers/handler.js (hunk 4 above)
Comment (ggrg) [reply src/handlers/transfers/handler.js]: ok

File: src/handlers/transfers/handler.js (hunk 5 above)
Comment (ggrg) [reply src/handlers/transfers/handler.js]: ok

File: src/handlers/transfers/handler.js (hunk 6 above)
Comment (ggrg) [reply src/handlers/transfers/handler.js]: ok

File: src/handlers/transfers/handler.js [hunk 7]
```diff
@@ -226,10 +219,13 @@ const fulfil = async (error, messages) => {
       histTimerEnd({ success: false, fspId: Config.INSTRUMENTATION_METRICS_LABELS.fspId })
       return true
     }
-    // Action Letter (AL) is used to denote breadcrumb flow end
-    const AL = action === TEA.COMMIT ? 'C' : (action === TEA.REJECT ? 'R' : (action === TEA.ABORT ? 'A' : '?'))
+    const CommitActionLetter = 'C'
```
Comment (rmothilal) [inline src/handlers/transfers/handler.js]: Should all of these not be put in an enum in the enum file?

File: src/handlers/transfers/handler.js [hunk 8]
```diff
@@ -120,8 +112,9 @@ const prepare = async (error, messages) => {
       histTimerEnd({ success: false, fspId: Config.INSTRUMENTATION_METRICS_LABELS.fspId })
       return true
     }
-    // Action Letter (AL) is used to denote breadcrumb flow end
-    const AL = action === TEA.PREPARE ? 'P' : '?'
+    const PrepareActionLetter = 'P'
```
Comment (rmothilal) [inline src/handlers/transfers/handler.js]: Should all of these not be put in an enum in the enum file?

File: src/handlers/transfers/handler.js (hunk 7 above)
Comment (ggrg) [reply src/handlers/transfers/handler.js]: yes, it makes sense

File: src/handlers/transfers/handler.js (hunk 8 above)
Comment (ggrg) [reply src/handlers/transfers/handler.js]: yes, it makes sense

File: src/domain/transfer/transform.js [hunk 9]
```diff
@@ -139,13 +130,13 @@ const toTransfer = (t) => {
...
     Logger.debug('In aggregate transfer transform for isTransferReadModel')
-    return Util.filterUndefined(fromTransferReadModel(t)) // TODO: Remove this once the DB validation is done for 't'
+    return Util.omitNil(fromTransferReadModel(t)) // TODO: Remove this once the DB validation is done for 't'
   } else if (t.isSaveTransferPrepared) {
     Logger.debug('In aggregate transfer transform for isSaveTransferPrepared')
-    return Util.filterUndefined(fromSaveTransferPrepared(t)) // TODO: Remove this once the DB validation is done for 't'
+    return Util.omitNil(fromSaveTransferPrepared(t)) // TODO: Remove this once the DB validation is done for 't'
   } else if (t.saveTransferFulfilledExecuted) {
     Logger.debug('In aggregate transfer transform for isSaveTransferExecuted')
-    return Util.filterUndefined(fromSaveTransferExecuted(t)) // TODO: Remove this once the DB validation is done for 't'
```
Comment (mdebarros) [inline src/domain/transfer/transform.js]: We should remove the Util.filterUndefined as the omitNil does the job already.

File: src/lib/errors.js [hunk 10]
```diff
@@ -24,9 +24,19 @@
 
 'use strict'
 
+const errorTitle = {
```
Comment (mdebarros) [inline src/lib/errors.js]: I would rather rename this to "errorType" as this is not a `title`.

File: src/domain/transfer/transform.js (hunk 9 above)
Comment (ggrg) [reply src/domain/transfer/transform.js]: Done. I thought we might have need it where we want to remove just undefined, not nulls.

File: src/lib/errors.js (hunk 10 above)
Comment (ggrg) [reply src/lib/errors.js]: Changed. Thanks for the review.

**Pull request merged**